proof/refine/RISCV64/KHeap_R.thy

```diff
@@ -207,64 +373,261 @@ lemma obj_at_setObject2:
+  shows "ko_wp_at' P ptr (set_ko' ptr' ko s) = ko_wp_at' P ptr s"
+  using assms
+  apply (clarsimp simp: ko_wp_at'_def)
+  apply (rule iffI; clarsimp)
+   apply (erule ps_clear_domE)
+   apply clarsimp
+   apply blast
+  apply (erule ps_clear_domE)
+  apply clarsimp
+  apply blast
+  done
 
-lemma updateObject_tcb_eta:
-  "updateObject (v :: tcb) = updateObject_default v"
-  by ((rule ext)+, simp)
+lemma obj_at'_set_obj'_distinct:
+  assumes "p \<noteq> p'"
+          "obj_at' Q p' s"
+  shows "obj_at' P p (set_ko' p' ko s) = obj_at' P p s"
+  using assms apply -
+  apply (rule iffI; fastforce simp: obj_at'_def projectKO_eq project_inject ps_clear_upd)
+  done
 
-lemma updateObject_ntfn_eta:
-  "updateObject (v :: Structures_H.notification) = updateObject_default v"
-  by ((rule ext)+, simp)
+lemmas pred_tcb_at'_set_obj'_distinct =
+  obj_at'_set_obj'_distinct[where P="test o proj o tcb_to_itcb'" for test proj,
+                            simplified o_def, folded pred_tcb_at'_def]
 
-lemmas updateObject_eta =
-  updateObject_ep_eta updateObject_tcb_eta updateObject_ntfn_eta
+lemmas pred_tcb_at'_set_obj'_iff =
+  tcb_obj_at'_set_obj'_iff[where Q="test o proj o tcb_to_itcb'" for test proj,
+                           simplified o_def injectKO_tcb, folded pred_tcb_at'_def]
 
-lemma objBits_type:
-  "koTypeOf k = koTypeOf k' \<Longrightarrow> objBitsKO k = objBitsKO k'"
-  by (erule koType_objBitsKO)
+lemma non_sc_same_typ_at'_objBits_always_the_same:
+  assumes "typ_at' t ptr s"
+          "koTypeOf ko = t"
+          "t \<noteq> SchedContextT"
+  shows "ko_wp_at' (\<lambda>old_ko. objBitsKO old_ko = objBitsKO ko) ptr s"
+  using assms
+  apply (clarsimp simp: typ_at'_def ko_wp_at'_def)
+  apply (rule koType_objBitsKO)
+  apply simp+
+  done
 
-lemma setObject_typ_at_inv:
-  "\<lbrace>typ_at' T p'\<rbrace> setObject p v \<lbrace>\<lambda>r. typ_at' T p'\<rbrace>"
-  apply (clarsimp simp: setObject_def split_def)
-  apply (clarsimp simp: valid_def typ_at'_def ko_wp_at'_def in_monad
-                        lookupAround2_char1 ps_clear_upd)
-  apply (drule updateObject_type)
-  apply clarsimp
-  apply (drule objBits_type)
-  apply (simp add: ps_clear_upd)
+lemmas non_sc_same_typ_at'_ko_wp_at'_set_ko'_iff =
+  same_size_ko_wp_at'_set_ko'_iff[OF non_sc_same_typ_at'_objBits_always_the_same]
+
+(* Worth adding other typ_at's? *)
+lemma typ_at'_ksPSpace_exI:
+  "pte_at' ptr s \<Longrightarrow> \<exists>pte. ksPSpace s ptr = Some (KOArch (KOPTE pte))"
+  apply -
+  apply (clarsimp simp: typ_at'_def ko_wp_at'_def,
+         (case_tac ko; clarsimp),
+         (rename_tac arch, case_tac arch; clarsimp)?)+
   done
 
+\<comment>\<open> Used to show a stronger variant of @{thm obj_at_setObject2} for many concrete types.
+
+    Needs to be a definition so we can easily refer to it within ML as a constant. \<close>
+definition distinct_updateObject_types ::
+  "('a :: pspace_storable) itself \<Rightarrow> ('b :: pspace_storable) itself \<Rightarrow> bool"
+where
```

Comment:
where should go on previous line (or be indented by 2 if it doesn't fit), so that `diff` and friends recognise the definition as one block.